### ISSUE TITLE
Microsoft Teams integration

### ISSUE THREAD

**Issue by olliem** [metabase/metabase#8051]
Hi all. First off, Metabase is amazing and our team loves it and uses it every single day.

I wanted to make a quick feature request as we are currently in the process of migrating from Slack over to Microsoft Teams. We use the Slack integration all time therefore its going to be a feature sadly missed without integration with Microsoft Teams.

Would it possible to consider adding support for Microsoft Teams?

Thanks!

**Comment by jornh** [comment]
Which part of the Slack integration is it you’re missing the most? If it’s subscription to alerts you may be able to substitute that by sending email alerts to the email address of a Teams channel (haven’t tried it myself — yet)

Interactive bot use of course requires more work ...

**Comment by olliem** [comment]
@jornh good idea - it's mainly alerts and pulses we're really interested in so this could defo work. I'll give it a try.

**Comment by omegadeep10** [comment]
Instead of another proprietary integration, do you think it would be a better idea to add a general "web hook" integration? As far as I know, most chat clients such as Rocket.Chat, Slack, Mattermost, etc. support web hooks. Not sure about Microsoft Teams.
@jornh

**Comment by jornh** [comment]
Good suggestions @omegadeep10 

* A pulse webhook #2298 has already been proposed. I guess all a webhook really does specify is the mechanism with a http POST but not the format of the content (though JSON is pretty common ;)). Would the content formatting proposed in #2298 work directly? For Teams? For the other desired endpoints? If not I guess it’s a matter of “stamping” that same content through a template. Should that be a capability of Metabase or through some external transformation?

* Regarding other uses a discussion is happening on #4120. It seems that discussion is really about (at least) two uses though: a) People who want to be able to be notified (on a schedule) of data from reported out of Metabase. b) People who want to subscribe to CRUD changes of Metabase objects like cards, dashboards (similar’ish to the activity feed on the front page of Metabase 0.29 - and tucked away under the 🔔 icon in 0.30-snapshot as of now)

**Comment by omegadeep10** [comment]
This is how healthchecks does flexible webhooks:

![image](https://user-images.githubusercontent.com/14183626/43264183-60dea182-90b3-11e8-82a0-091ad5f1ef56.png)

The core idea here is that the system provides variables the user might be interested in capturing, and then it is up to the user to build the actual webhook by specifying endpoint, JSON body, request headers, etc. This allows the user to build a webhook for any possible integration, since everyone does webhooks a differently.

Usability wise, is this good for Metabase? I think that depends on the type of user. I personally think that this is very straightforward and easy to use if you have a basic understanding of an HTTP request and JSON structure. However, I understand that most Metabase users, the business users, will not be able to understand and use this feature effectively. 

I think it's safe to assume Administrators have technical knowledge (especially since other features in the Admin Panel such as LDAP or email require some technical knowledge). How about instead, we let Administrators build generalized webhook "templates" that can be easily used by the business user?

Example:
- User Joe wants to receive pulse notifications in the Rocket.Chat #sales channel.

Let's assume no webhook "templates" exist yet. Joe should ask Admin Bob to build one. Admin Bob goes into the Admin panel, and navigates to the "webhooks" section. Admin Bob selects the type of webhook he wants to build.

The types of webhooks correspond to the different types of events. E.g. pulse, card change, etc.

Admin Bob selects "Pulse". A simple UI appears similar to above that has pulse specific variables, such as pulse title, pulse author, links to the pulse data, etc. Admin Bob builds the webhook by specifying endpoint, JSON body with variables within the body, and any relevant headers. Bob should also name the webhook something descriptive, like "Post to #sales"

Once saved, User Joe can now create a pulse and specify the endpoint "Post to #sales" which corresponds to Admin Bob's webhook template. The webhook option might appear as another option below the standard email and slack options:

![image](https://user-images.githubusercontent.com/14183626/43265762-b78a05b8-90b7-11e8-9e37-147983d4de17.png)


I think this is a good compromise of flexibility and usability. End users don't have to know anything about Webhooks to use them, while Admins only need to build the webhook "template" once.

I do think there are further questions to be answered regarding things like customizable webhooks for example. What if Joe wanted to send the pulse as a direct message to himself? It would be infeasible to expect Admin Bob to build a webhook template for every user that wanted to send a pulse as a direct message to themselves. Perhaps the user can pass variables that affect the webhook endpoint/JSON body? I'm not sure.

Also, how would other webhook events be handled? UI decisions would need to be made when deciding where to add the option so the user can configure a webhook for activity feed, or card update or something similar.

Thoughts? @jornh

**Comment by jornh** [comment]
@olliem it's been a while since you, regarding hooking up email notifications to MS Teams, said:

> good idea - it's mainly alerts and pulses we're really interested in so this could defo work. I'll give it a try.

Curious to know if you had a chance to try it out - and what your experiences have been? I guess mainly does the emails render OK in Teams or not?

**Comment by matheusjerico** [comment]
up

**Comment by orionich** [comment]
Yes, need webhooks integration. It's universal.

**Comment by flamber** [comment]
@orionich There's a request for webhooks specifically - #2298 - upvote by clicking :+1: on the first post

**Comment by bartoszhernas** [comment]
I've just used an email address to send to a channel on Microsoft Teams. Works brilliantly, on par with official Slack integration as the whole email is properly displayed

**Comment by thomasLeclaire** [comment]
Hello! we used the send to a channel for several years now and was perfectly working, but it sounds broken since a release on last monthes (didn't identified exactly which one for now..). I can confirm mail is sent correctly, so it's not really a Metabase issue, but sounds there are some incompatibility with the expected/allowed format on Teams side. 
I tried several smtp servers with same result. 

However, I can receive the alert on my personal mail and if I forward to the channel mail, it's working (but display is broken) which tend to confirm some incompatibility somewhere.
